packages/lesswrong/server/dialogues/cron.ts

```diff
@@ -21,3 +24,81 @@ addCronJob({
     })
   }
 });
+
+const checkCompatible = (preferences:DbDialogueMatchPreference[]) => {
+  if (preferences.length >= 2) {
+
+    // Topic
+    const sourceUserYesTopics = getUserTopics(preferences[0]);
+    const targetUserYesTopics = getUserTopics(preferences[1]);
+    const sharedTopics = sourceUserYesTopics.filter(topic => targetUserYesTopics.includes(topic));
+
+    // Format 
+    const { sync: userSync, async: userAsync } = getSyncAsyncPreferences(preferences[0]);
+    const { sync: targetUserSync, async: targetUserAsync } = getSyncAsyncPreferences(preferences[1]);
+    const { formatPreferenceMatch } = getPreferenceMatches(
+      { sync: userSync, async: userAsync },
+      { sync: targetUserSync, async: targetUserAsync }
+    );
+
+    if (sharedTopics.length > 0 && formatPreferenceMatch) {
+      return true
+    } 
+  } 
+  return false
+}
+
+addCronJob({
+  name: 'sendDialogueHelperBotPing',
+  interval: 'every 30 seconds', // `every 12 hours`,
+  async job() {
+    const context = createAdminContext();
+
+    const staleDialogues = await context.repos.posts.getStaleDialogues()
+    // filter for compatible (format and topic match)
+    const compatibleDialogues = [];
+    for (const dialogue of staleDialogues) {
+      const dialogueMatchPreferences = await DialogueMatchPreferences.find({
+        generatedDialogueId: dialogue._id,
+        deleted: false
+      }).fetch();
+
+      const isCompatible = checkCompatible(dialogueMatchPreferences); // todo: debug
+      if (isCompatible) {
+        compatibleDialogues.push(dialogue);
+      }
+    }
```

Comment:
Pretty sure you can do this in a single query (rather than iterating over `staleDialogues`, of which there are >100 in prod right now):
```
const staleDialogueIds = staleDialogues.map(({ _id }) => _id);
const dialogueMatchPreferences = await DialogueMatchPreferences.find({
  generatedDialogueId: { $in: staleDialogueIds },
  deleted: false
}).fetch();

const matchPreferencesByDialogueId = groupBy(dialogueMatchPreferences, (matchPreference) => matchPreference.generatedDialogueId);

const compatibleMatchPreferenceDialogueIds = new Set(Object.keys(matchPreferencesByDialogueId).filter(dialogueId => isCompatible(matchPreferencesByDialogueId[dialogueId])));

const compatibleDialogues = staleDialogues.filter(staleDialogue => compatibleMatchPreferenceDialogueIds.has(staleDialogue));
```

Reply:
(except the `compatibleMatchPreferenceDialogueIds` declaration should probably be broken up into two lines, or something more readable)